src/sources/file.rs

```diff
@@ -193,6 +224,86 @@ pub fn file_source(
     }))
 }
 
+struct LineAgg<T> {
+    inner: T,
+    marker: Regex,
+    timeout: u64,
+    buffers: HashMap<String, Bytes>,
+    draining: Option<Vec<(Bytes, String)>>,
+    timeouts: DelayQueue<String>,
+}
+
+impl<T> LineAgg<T> {
+    fn new(inner: T, marker: Regex, timeout: u64) -> Self {
+        Self {
+            inner,
+            marker,
+            timeout,
+            draining: None,
+            buffers: HashMap::new(),
+            timeouts: DelayQueue::new(),
+        }
+    }
+}
+
+impl<T: Stream<Item = (Bytes, String), Error = ()>> Stream for LineAgg<T> {
+    type Item = (Bytes, String);
+    type Error = ();
+
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        loop {
+            if let Some(to_drain) = &mut self.draining {
+                if let Some((data, key)) = to_drain.pop() {
+                    return Ok(Async::Ready(Some((data, key))));
+                } else {
+                    return Ok(Async::Ready(None));
+                }
+            }
+
+            match self.inner.poll() {
+                Ok(Async::Ready(Some((line, src)))) => {
+                    // look for buffered content from same source
+                    if let Some(buffered) = self.buffers.remove(&src) {
+                        if self.marker.is_match(line.as_ref()) {
+                            // buffer the incoming line and flush the existing data
+                            self.buffers.insert(src.clone(), line);
+                            return Ok(Async::Ready(Some((buffered, src))));
+                        } else {
+                            // append new line to the buffered data
+                            let mut more = buffered.to_vec();
```

Comment:
I think it would be better to use `BytesMut` within the map here then use `BufMut` items to extend the buffered bytes. Then once you're ready to return the multi-line event, just call `BytesMut::freeze` which should then convert the `BytesMut` into `Bytes`. This should allow us to avoid this allocation of a vec on every line that doesnt match.

This means, I think we should check the match of the line first, then call remove or get_mut. This should clean up this block and reduce the amount of removes & inserts.

Reply:
Good call on `BytesMut`, I meant to go back and do that.

I'm less sold on the second half. Other than fewer removes/inserts, this doesn't seem nicer:
```diff
diff --git a/src/sources/file.rs b/src/sources/file.rs
index 17bf049..de87d1e 100644
--- a/src/sources/file.rs
+++ b/src/sources/file.rs
@@ -263,16 +263,22 @@ impl<T: Stream<Item = (Bytes, String), Error = ()>> Stream for LineAgg<T> {
             match self.inner.poll() {
                 Ok(Async::Ready(Some((line, src)))) => {
                     // look for buffered content from same source
-                    if let Some(mut buffered) = self.buffers.remove(&src) {
+                    if self.buffers.contains_key(&src) {
                         if self.marker.is_match(line.as_ref()) {
                             // buffer the incoming line and flush the existing data
-                            self.buffers.insert(src.clone(), line.into());
+                            let buffered = self
+                                .buffers
+                                .insert(src.clone(), line.into())
+                                .expect("already asserted key is present");
                             return Ok(Async::Ready(Some((buffered.freeze(), src))));
                         } else {
                             // append new line to the buffered data
+                            let buffered = self
+                                .buffers
+                                .get_mut(&src)
+                                .expect("already asserted key is present");
                             buffered.extend_from_slice(b"\n");
                             buffered.extend_from_slice(&line);
-                            self.buffers.insert(src, buffered);
                         }
                     } else {
```
Is that what you had in mind?

Reply:
Yup, this is much better.

Reply:
You prefer to include the diff I put in my previous comment? It feels a little noisy with the redundant `contains_key` and `insert().expect()`/`get_mut().expect()`, but I can see the argument.

Reply:
I think it makes it a bit more clear on intentions. Up to you 😄